src/psbt.c

```diff
@@ -2201,3 +2203,752 @@ int wally_psbt_to_base64(
+            }
+        }
+    }
+
+    *output = result;
+    return WALLY_OK;
+
+fail:
+    if (result) {
+        wally_psbt_free(result);
+    }
+    return ret;
+}
+
+int wally_sign_psbt(
+    struct wally_psbt *psbt,
+    const unsigned char *key,
+    size_t key_len)
+{
+    unsigned char pubkey[EC_PUBLIC_KEY_UNCOMPRESSED_LEN], uncomp_pubkey[EC_PUBLIC_KEY_UNCOMPRESSED_LEN], sig[EC_SIGNATURE_LEN], der_sig[EC_SIGNATURE_DER_MAX_LEN + 1];
+    size_t i, j, der_sig_len;
+    int ret;
+
+    if (!psbt || !key) {
+        return WALLY_EINVAL;
+    }
+
+    // Get the pubkey
+    if ((ret = wally_ec_public_key_from_private_key(key, key_len, pubkey, EC_PUBLIC_KEY_LEN)) != WALLY_OK) {
+        return ret;
+    }
+    if ((ret = wally_ec_public_key_decompress(pubkey, EC_PUBLIC_KEY_LEN, uncomp_pubkey, EC_PUBLIC_KEY_UNCOMPRESSED_LEN)) != WALLY_OK) {
+        return ret;
+    }
+
+    // Go through each of the inputs
+    for (i = 0; i < psbt->num_inputs; ++i) {
+        struct wally_psbt_input *input = &psbt->inputs[i];
+        struct wally_tx_input *txin = &psbt->tx->inputs[i];
+        unsigned char sighash[SHA256_LEN], *scriptcode, wpkh_sc[WALLY_SCRIPTPUBKEY_P2PKH_LEN];
+        size_t scriptcode_len;
+        bool match = false, comp = false;
+        uint32_t sighash_type = WALLY_SIGHASH_ALL;
+
+        if (!input->keypaths) {
+            // Can't do anything without the keypaths
+            continue;
+        }
+
+        // Go through each listed pubkey and see if it matches.
+        for (j = 0; j < input->keypaths->num_items; ++j) {
+            struct wally_keypath_item *item = &input->keypaths->items[j];
+            if (item->pubkey[0] == 0x04 && memcmp((char *)item->pubkey, (char *)uncomp_pubkey, EC_PUBLIC_KEY_UNCOMPRESSED_LEN) == 0) {
+                match = true;
+                break;
+            } else if (memcmp((char *)item->pubkey, (char *)pubkey, EC_PUBLIC_KEY_LEN) == 0) {
+                match = true;
+                comp = true;
+                break;
+            }
+        }
+
+        // Did not find pubkey, skip
+        if (!match) {
+            continue;
+        }
+
+        // Sighash type
+        if (input->sighash_type > 0) {
+            sighash_type = input->sighash_type;
+        }
+
+        // Get scriptcode and sighash
+        if (input->redeem_script) {
+            unsigned char sh[WALLY_SCRIPTPUBKEY_P2SH_LEN];
+            size_t written;
+
+            if ((ret = wally_scriptpubkey_p2sh_from_bytes(input->redeem_script, input->redeem_script_len, WALLY_SCRIPT_HASH160, sh, WALLY_SCRIPTPUBKEY_P2SH_LEN, &written)) != WALLY_OK) {
+                return ret;
+            }
+            if (input->non_witness_utxo) {
+                if (input->non_witness_utxo->outputs[txin->index].script_len != WALLY_SCRIPTPUBKEY_P2SH_LEN ||
+                    memcmp((char *)sh, (char *)input->non_witness_utxo->outputs[txin->index].script, WALLY_SCRIPTPUBKEY_P2SH_LEN) != 0) {
+                    return WALLY_EINVAL;
```

Comment:
WALLY_ERROR

Reply:
I think EINVAL fits better here as it's checking that the psbt fields are consistent with each other.